### PR TITLE
add out option to cupy.core.core.scan & performance improvement

### PR DESCRIPTION
**Pull request by kikusu**
This PR adds out option to cupy.core.core.scan.

Performance improvement
```
# before
In [2]: a = cupy.arange(10000)
In [3]: %timeit cupy.core.core.scan(a)
The slowest run took 143.12 times longer than the fastest. This could mean that an intermediate result is being cached.
1000 loops, best of 3: 259 µs per loop

# after
In [2]: a = cupy.arange(10000)
In [3]: %timeit cupy.core.core.scan(a)
The slowest run took 13.74 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 35.7 µs per loo

In [4]: b = cupy.zeros_like(a)
In [5]: %timeit cupy.core.core.scan(a,b)
The slowest run took 25.18 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 18.7 µs per loop
```

### CODE REVIEW

**Comment by kikusu** [comment]
The bug was found, so it's corrected.

**Comment by kikusu** [comment]
I fixed bug.

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -2416,14 +2416,14 @@ def _add_scan_blocked_sum_kernel(dtype):
...
     dtype = _get_typename(dtype)
     source = string.Template("""
-    extern "C" __global__ void ${name}(CArray<${dtype}, 1> src_dst,
-        CArray<${dtype}, 1> sum){
+    extern "C" __global__ void ${name}(CArray<${dtype}, 1> src_dst){
         long long n = src_dst.size();
-        unsigned int idx = threadIdx.x + blockDim.x * blockIdx.x;
-        unsigned int idxSum = idx + blockDim.x;
+        unsigned int idx = threadIdx.x + (blockDim.x + 1) * blockIdx.x;
+        unsigned int idxAdded = idx + (blockDim.x + 1);
```
Comment (okuta) [inline cupy/core/core.pyx]: How about ``unsigned int idxAdded = threadIdx.x + (blockDim.x + 1) * (blockIdx.x + 1);``?

**Comment by okuta** [comment]
LGTM except comment.

File: cupy/core/core.pyx (hunk 1 above)
Comment (kikusu) [inline cupy/core/core.pyx]: Thank you for review. I fixed it.

**Comment by okuta** [comment]
Thank you!

**Pull request merged**